core/src/main/java/org/hisp/dhis/android/core/constant/internal/ConstantModuleDownloader.kt

```diff
@@ -25,30 +25,23 @@
  *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
  *  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
-package org.hisp.dhis.android.core.constant.internal;
+package org.hisp.dhis.android.core.constant.internal
 
-import org.hisp.dhis.android.core.arch.modules.internal.TypedModuleDownloader;
-import org.hisp.dhis.android.core.constant.Constant;
-
-import java.util.List;
-
-import javax.inject.Inject;
-
-import dagger.Reusable;
-import io.reactivex.Single;
+import dagger.Reusable
+import io.reactivex.Single
+import javax.inject.Inject
+import kotlinx.coroutines.rx2.rxSingle
+import org.hisp.dhis.android.core.arch.modules.internal.TypedModuleDownloader
+import org.hisp.dhis.android.core.constant.Constant
 
 @Reusable
-public class ConstantModuleDownloader implements TypedModuleDownloader<List<Constant>> {
-
-    private final ConstantCallFactory constantCallFactory;
-
-    @Inject
-    ConstantModuleDownloader(ConstantCallFactory constantCallFactory) {
-        this.constantCallFactory = constantCallFactory;
-    }
-
-    @Override
-    public Single<List<Constant>> downloadMetadata() {
-        return Single.fromCallable(constantCallFactory.create());
+class ConstantModuleDownloader @Inject internal constructor(
+    private val constantCallFactory: ConstantCoroutineCallFactory
+) :
+    TypedModuleDownloader<List<Constant>> {
+    override fun downloadMetadata(): Single<List<Constant>> {
```

Comment:
It seems this class could be marked as `internal`, and this method could return a suspend function instead of a `Single`. The class `MetadataCall` is ready to accept suspend functions in their methods.

Reply:
Yes it could be internal, but for using coroutines, I think we either have to remove `TypeModuleDownloader` or refactor it to Coroutines (it's used by a couple more classes)

Reply:
It could be duplicated if we don't want to migrate all the usages of `TypeModuleDownloader`

Reply:
I'd refactor it to use coroutines, so we can have one more thing migrated